### ISSUE TITLE
Add event hook for cancelable browser back and forward navigation

### ISSUE THREAD

**Issue by Bloemendaal** [inertiajs/inertia#1645]
### Versions:

- `@inertiajs/core` version: 1.0.10

### Describe the problem:

Currently, Inertia does not provide a specific event hook for intercepting and canceling browser back and forward navigation. While the library allows capturing in-page navigation, handling navigation through the browser's back and forward buttons is not directly supported. This limitation poses challenges for scenarios such as warning users about unsaved changes, as well as other contexts where controlling or canceling back and forward navigation is necessary.

The expected behavior would be to have a way within Inertia to easily detect and optionally cancel back and forward navigation, possibly through specific events or hooks designed for this purpose.

### Steps to reproduce:

1. **Create a component with unsaved changes state**: Create a form with some input fields, and use a state to track whether there are unsaved changes.
2. **Modify without saving**: Modify the input without saving, thus setting the unsaved changes state to true.
3. **Attempt to navigate away**: Use the browser's back or forward buttons to attempt navigation away from the page.
4. **No handling for unsaved changes**: Observe that there is no built-in way within Inertia to intercept this navigation, warn the user, or cancel the navigation.

The absence of an event hook or mechanism within Inertia to handle browser back and forward navigation necessitates workarounds, such as using the `popstate` event, which may interfere with other client-side routing logic.

### Preferred solution
```ts
export default function useUnsavedChanges(active: boolean): void {
  useEffect(() => {
    const handleInertiaNavigation = (event: Event): void => {
      if (hasUnsavedChanges) {
        const confirmed = window.confirm('You have unsaved changes! Are you sure you want to leave?');

        if (!confirmed) {
          event.preventDefault();
        }
      }
    };

    return router.on('beforeNavigateHistory', handleInertiaNavigation);
  }, [hasUnsavedChanges]);
}
```

**Comment by RobertBoes** [comment]
Correct me if I'm wrong, but cancelling the history navigation is impossible, otherwise it would be incredibly easy to trap someone on your website. There is the `onbeforeunload` event, which is cancellable by the user, maybe the same could be done here (I'm talking about the native browser `confirm` popup). Apart from that I don't think there would be any way to cancel a history navigation, not without hacky workarounds such as pushing a new state, but many browsers block that too afaik

**Comment by Bloemendaal** [comment]
You're right, but according to this 'very reliable' source, pushing the same state back should work: https://chat.openai.com/share/00d4058e-bbd1-46eb-8543-95a436bf613f

At least I'd like to be able to cancel the functions Inertia executes with this event to implement my own hack. Normally, the browser would dispatch an unload event when using normal navigation. So, I think it should be possible to replicate the same functionality.

I'm building an editor with react flow and if the user moves left, the native back even in the browser is sometimes triggered by accident and the complete state is gone :/
